### PR TITLE
Security

### LINKED ISSUE #35: Enable security
First part of Security.
https://www.youtube.com/watch?v=5QEtVG0Dh1I&list=PLVApX3evDwJ1d0lKKHssPQvzv2Ao3e__Q&index=9

### LINKED ISSUE #36: HTTP basic Authentication
Second part of security
https://www.youtube.com/watch?v=hF-iMHpl970&list=PLVApX3evDwJ1d0lKKHssPQvzv2Ao3e__Q&index=11

### LINKED ISSUE #37: Configure Role Based Authorization
Thired part of security
https://www.youtube.com/watch?v=tYpS8BmrEyo&list=PLVApX3evDwJ1d0lKKHssPQvzv2Ao3e__Q&index=12

### LINKED ISSUE #38: Secure REST controllers
Forth part of security
https://www.youtube.com/watch?v=OYr9HUPmhSw&list=PLVApX3evDwJ1d0lKKHssPQvzv2Ao3e__Q&index=13

### LINKED ISSUE #40: Database Authentication
Sixed (and last) part of security.

Video 16-20 (start: https://www.youtube.com/watch?v=4l5d-ZVkDPg&list=PLVApX3evDwJ1d0lKKHssPQvzv2Ao3e__Q&index=16)

Database authentication contations the following steps:
- Create a user entitet
- Store said user in our database
- Link our user entity to the build in classes in spring security
- Integrate ”Database Auth” in our configuration

### PR DESCRIPTION
**Pull request by helenahalldiniths**
This commit will fix #35, fix #36, fix #37, fix #38, fix #40

### CODE REVIEW

**Review by FelixJacobsen — APPROVED** [review]
I think this is a good start implementation for Spring security.

In the future we might wanna consider adding: 
1. A form based security with a Thymeleaf interface
2. Seperate security entities like the following example: https://www.baeldung.com/role-and-privilege-for-spring-security-registration
3. A implementation of a token-system 

Great job!

File: src/main/java/se/iths/crimedatabase/security/SecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        return PasswordEncoderFactories.createDelegatingPasswordEncoder();
+    }
+
+    @Bean
+    public DaoAuthenticationProvider authenticationProvider() {
+        DaoAuthenticationProvider provider = new DaoAuthenticationProvider();
+        provider.setPasswordEncoder(passwordEncoder());
+        provider.setUserDetailsService(userPrincipalDetailsService);
+        return provider;
+    }
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/security/SecurityConfig.java:54]:
```suggestion
```

Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/security/SecurityConfig.java]:
```suggestion
```

File: src/main/java/se/iths/crimedatabase/security/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    public SecurityConfig(UserPrincipalDetailsService userPrincipalDetailsService) {
+        this.userPrincipalDetailsService = userPrincipalDetailsService;
+    }
+
+    //Defines datasource for our users
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) {
+        auth.authenticationProvider(authenticationProvider());
+    }
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/security/SecurityConfig.java]:
```suggestion
```

File: src/main/java/se/iths/crimedatabase/security/UserPrincipal.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+import se.iths.crimedatabase.entity.User;
+
+import java.util.Collection;
+import java.util.List;
+
+public class UserPrincipal implements UserDetails {
+    private User user;
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/security/UserPrincipal.java]:
```suggestion
    private final User user;
```

File: src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    public UserPrincipalDetailsService(UserRepository userRepository) {
+        this.userRepository = userRepository;
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        Optional<User> user = Optional.ofNullable(userRepository.findByUsername(username));
+        if (user.isEmpty())
+            throw new UsernameNotFoundException("Can't find user with username " + username);
+        return new UserPrincipal(user.get());
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java:25]:
```suggestion
        User user = userRepository.findByUsername(username);
        if (user == null)
            throw new UsernameNotFoundException("Can't find user with username " + username);
        return new UserPrincipal(user);
```

File: src/main/java/se/iths/crimedatabase/security/UserInit.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import org.springframework.stereotype.Service;
+import se.iths.crimedatabase.entity.User;
+import se.iths.crimedatabase.repository.UserRepository;
+
+import java.util.List;
+
+@Service
+public class UserInit implements CommandLineRunner {
+    private UserRepository userRepository;
+    private PasswordEncoder passwordEncoder;
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/security/UserInit.java]:
```suggestion
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
```

File: src/main/java/se/iths/crimedatabase/security/SecurityConfig.java (hunk 1 above)
Comment (LordRekishi) [reply src/main/java/se/iths/crimedatabase/security/SecurityConfig.java]: We do not need this it seems

File: src/main/java/se/iths/crimedatabase/security/SecurityConfig.java (hunk 2 above)
Comment (LordRekishi) [reply src/main/java/se/iths/crimedatabase/security/SecurityConfig.java]: We don't need this it seems

File: src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+import se.iths.crimedatabase.entity.User;
+import se.iths.crimedatabase.repository.UserRepository;
+
+import java.util.Optional;
+
+@Service
+public class UserPrincipalDetailsService implements UserDetailsService {
+    UserRepository userRepository;
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java]:
```suggestion
    private final UserRepository userRepository;
```

File: src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java (hunk 4 above)
Comment (LordRekishi) [reply src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java:25]: This is the way we made the method during Martin's class. For/against?

**Review by LordRekishi — CHANGES_REQUESTED** [review]
Please take a look at my changes. Everything has been tested locally and works.

File: src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java (hunk 4 above)
Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java:25]: I think we should avoid using null!

File: src/main/java/se/iths/crimedatabase/security/SecurityConfig.java (hunk 2 above)
Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/security/SecurityConfig.java]: Good! I will do some tests and remove it if all looks good!

File: src/main/java/se/iths/crimedatabase/security/SecurityConfig.java (hunk 1 above)
Comment (helenahalldiniths) [reply src/main/java/se/iths/crimedatabase/security/SecurityConfig.java]: Good! I will do some tests and remove it if all looks good!

Comment (LordRekishi) [reply src/main/java/se/iths/crimedatabase/security/SecurityConfig.java]: Remember to remove both this and the next sections from the SecurityConfig file before testing :)

File: src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java (hunk 4 above)
Comment (LordRekishi) [reply src/main/java/se/iths/crimedatabase/security/UserPrincipalDetailsService.java:25]: Fair enough!

**Comment by helenahalldiniths** [comment]
I agree with all of your suggestions @LordRekishi :)

File: src/main/java/se/iths/crimedatabase/entity/User.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public String getPassword() {
+        return password;
+    }
+
+    public String getRole() {
+        return role;
+    }
+}
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/entity/User.java]:
```suggestion
}

```

File: src/main/java/se/iths/crimedatabase/controller/UserController.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+import se.iths.crimedatabase.entity.User;
+import se.iths.crimedatabase.repository.UserRepository;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("users")
+public class UserController {
+
+    UserRepository userRepository;
```
Comment (LordRekishi) [inline src/main/java/se/iths/crimedatabase/controller/UserController.java]:
```suggestion
    private final UserRepository userRepository;
```

**Review by LordRekishi — CHANGES_REQUESTED** [review]

**Comment by LordRekishi** [comment]
> I agree with all of your suggestions @LordRekishi :)

I made some more :D

**Review by LordRekishi — APPROVED** [review]

**Pull request merged**